projects/plugins/jetpack/class.jetpack-gutenberg.php

```diff
@@ -622,6 +622,18 @@ public static function enqueue_block_editor_assets() {
 			return;
 		}
 
+		/**
+		 * This can be called multiple times per page load in the admin, during the `enqueue_block_assets` action.
+		 * These assets are necessary for the admin for editing but are not necessary for each pattern preview.
+		 * Therefore we dequeue them, so they don't load for each pattern preview iframe.
+		 */
+		if ( ! wp_should_load_block_editor_scripts_and_styles() ) {
+			wp_dequeue_script( 'jp-tracks' );
+			wp_dequeue_script( 'jetpack-blocks-editor' );
```

Comment:
Two questions about this: why do we need to dequeue the two scripts? They are not going to be enqueued in this `enqueue_block_editor_assets` method. Then who else has already enqueued them?

For the iframed assets, the `_wp_get_iframed_editor_assets` is working on a private temporary instance of `WP_Scripts`, which only copies _registered_ scripts from the global instance, but not the _queued_ ones.

Second question: this method is called also in an non-admin context, see the condition here:

https://github.com/Automattic/jetpack/blob/42aec0a2c7b092bf2fb2c8d6fdca1aba428f457c/projects/plugins/jetpack/class.jetpack.php#L721

Apparently we want to enqueue the `jetpack-blocks-editor` script in some context that has `is_admin() == false`. Probably some site editor iframe? I can't be frontend because the `enqueue_block_editor_assets` is executed only in admin. Is it ever called at all?

If the intent is to enqueue `jetpack-blocks-editor` in some page where `is_admin()` is `false`, then your patch is preventing that because the `wp_should_load_block_editor_scripts_and_styles` will check `$current_screen->is_block_editor()`.

Reply:
The condition mentioned above was added in #32706 by @BogdanUngureanu. Maybe he knows what the non-admin code path does. And what exactly did that PR fix? The PR description mentions something about correctly rendering the Donations block, but without specific details.

Reply:
> why do we need to dequeue the two scripts? They are not going to be enqueued in this enqueue_block_editor_assets method. Then who else has already enqueued them?

They will be enqueued on the first call of `enqueue_block_editor_assets()`, which is part of the initial editor load, so when that method runs again (during subsequent `enqueue_block_editor_assets()` calls, like for retrieving the editor assets), they'll already be enqueued. But we don't need them, because they only add extra unused scripts that clutter network requests and block the pattern preview loading process. That's why we're dequeuing them. You can refer to my previous comment for more information on how this works: https://github.com/Automattic/jetpack/pull/34268#issuecomment-1824748005

> Apparently we want to enqueue the jetpack-blocks-editor script in some context that has is_admin() == false. Probably some site editor iframe? I can't be frontend because the enqueue_block_editor_assets is executed only in admin. Is it ever called at all?
> The condition mentioned above was added in #32706 by @BogdanUngureanu. Maybe he knows what the non-admin code path does. And what exactly did that PR fix? The PR description mentions something about correctly rendering the Donations block, but without specific details.

I'm not sure about that usage, too. As you mentioned it would be good to get some information from the folks who introduced it. 

> If the intent is to enqueue jetpack-blocks-editor in some page where is_admin() is false, then your patch is preventing that because the wp_should_load_block_editor_scripts_and_styles will check $current_screen->is_block_editor().

Which is fine by design, and if that's the case, then that code should also use the built-in filter of `wp_should_load_block_editor_scripts_and_styles()` to override the default behavior and allow loading of the editor assets outside of the admin context for that particular case. I'd be happy to add that where applicable, if we get a confirmation from the authors. 

Perhaps someone else from @Automattic/t-rex will know?

Reply:
@BogdanUngureanu thanks for elaborating! The approach appears to be causing performance issues for the pattern previews on WP.com though. Can you help us with some testing instructions on how to test the scenarios you described, so we can ensure we're not breaking something?

Reply:
Sure! :)

- Open the Site Editor and then the Global Styles panel
- Go to Typography > Text > Size input.
- Toggle the input to a slider (the icon in the right)
- Move the slider left and right
- You'll notice that the slider UI responds slower before the change I've introduced

You can see a difference in response time in these two videos:

https://github.com/Automattic/jetpack/assets/995955/e35582a4-ad88-4836-82dc-2896507b85ea


https://github.com/Automattic/jetpack/assets/995955/d579ef1e-5d19-4075-98bd-ee18a61d84d6

In the video there are different types of sliders, for the one I've mentioned above, the difference should be more visible because the inputs change faster with a regular slider.

Reply:
> They will be enqueued on the first call of `enqueue_block_editor_assets()`, which is part of the initial editor load, so when that method runs again (during subsequent `enqueue_block_editor_assets()` calls, like for retrieving the editor assets), they'll already be enqueued.

I think it works a bit differently: the first call of `enqueue_block_editor_assets` will enqueue the assets for the initial HTML page of the editor.

Second time it runs it is inside `_wp_get_iframed_editor_assets`. There the global `$wp_scripts` object has been replaced temporarily with a newly created instance, where nothing is enqueued. I verified on a simple site that all I need to do is `if ( ! wp_should_load() ) { return; }`. No need to dequeue `jetpack-blocks-editor` because it's not enqueued. Unless there is a _second_ `enqueue_block_assets` registered filter that will add it. Then the best fix is to find it and add the same condition there, too.

Reply:
> However, using the same enqueue hook on the front-end (e.g. p2 post comments) the assets weren't loaded properly.

Oh, so here we have a case where we want to enqueue `jetpack-blocks-editor` even on the frontend, in a non-admin environment. So that the P2 editor has Jetpack blocks available.

If this is true, and the P2 editor relies on this Jetpack code to load Jetpack blocks, then the current patch breaks P2. Jetpack blocks won't be available there. I didn't test this yet, but it's something to verify before shipping.

Reply:
Verified that P2 works with this patch. The `Isolated_Block_Editor` class (`plugins/p2/.../class-editor-isolated.php`) will load the `jetpack-blocks-editor` script. Because it calls `do_action( 'enqueue_block_editor_assets' )`, and part of that action is enqueueing all scripts that were found in `block.json` files, as the `editorScript` fields. And Jetpack blocks declare `jetpack-blocks-editor` this way.

Reply:
I don't think it will break Jetpack blocks in p2, because we're intentionally enabling that same filter there:

https://github.com/Automattic/p2/blob/9c8502cc17775e863450e6e42c9e720dff096dc8/plugins/p2/src/server/core/editor/class-editor-isolated.php#L15C55-L15C55

which will result in `wp_should_load_block_editor_scripts_and_styles()` returning true as well, ultimately keeping the Jetpack block editor assets still enqueued. 

I'll make sure to test with p2 though, just in case.

Reply:
> Sure! :)

Thanks a bunch for the instructions and videos @BogdanUngureanu! I'll make sure to test those cases and to make sure we don't reintroduce that performance bottleneck.

Reply:
The "if is admin then `enqueue_block_assets` otherwise `enqueue_block_editor_assets`" condition is present also in many themes shipped by Core, like here:

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwenty/functions.php#L440-L444

But it also has an extra "WordPress version is 6.3 or newer" condition. So, something about the behavior of these two actions must have changed in 6.3. @BogdanUngureanu Maybe Jetpack wants to check the 6.3 condition, too? I think it's worth investigating.

Reply:
@jeherve please correct me if I'm wrong, but IIRC backward compatibility policies of Jetpack with WP core were "current and previous major versions", right? While for themes (and especially core themes), it's a bigger deal, since I expect we strive to support older versions of WordPress.

So, if that's the backward compatibility support for Jetpack, and I understand this correctly, we shouldn't be concerned about WP older than 6.3.

Reply:
> we shouldn't be concerned about WP older than 6.3.

That's correct. Jetpack requires WordPress 6.3 and newer since #34127.

Reply:
> Open the Site Editor and then the Global Styles panel
> Go to Typography > Text > Size input.
> Toggle the input to a slider (the icon in the right)
> Move the slider left and right
> You'll notice that the slider UI responds slower before the change I've introduced
> You can see a difference in response time in these two videos:
> 
> In the video there are different types of sliders, for the one I've mentioned above, the difference should be more visible because the inputs change faster with a regular slider.

Thanks again for the videos and instructions, @BogdanUngureanu, I've just verified that the performance of various sliders in global styles is fine with this patch. 👍